### PR TITLE
[shard prop] interpolate

### PR DESCRIPTION
**Pull request by pianpwk**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #177471
* #177619
* #177618
* #176992
* __->__ #176991

```
(pytorch-3048) [pianpwk@devvm3048.dkl0 /data/users/pianpwk/ptclone/pytorch (b1582add)]$ python -m torch.distributed.tensor._ops.strategy_validation --op "_upsample_bilinear2d_aa,grid_sampler_2d,max_pool2d_with_indices_backward,nn.functional.adaptive_avg_pool1d,nn.functional.adaptive_avg_pool2d,nn.functional.grid_sample,nn.functional.interpolate,nn.functional.max_unpool1d,nn.functional.max_unpool2d,nn.functional.max_unpool3d" --max-samples 10 --allow-composite

  ======================================================================                                                                           
  Summary
  ======================================================================                                                                           
  Op                                     Correct  Incorrect  Missing    Time                                                                       
  --------------------------------------------------------------------------
  aten._upsample_bilinear2d_aa                21          0        0     7.6s                                                                      
  aten.grid_sampler_2d                        30          0        3   118.0s                                                                      
  aten.max_pool2d_with_indices_backward      164          0       28   ~8s                                                                         
  nn.functional.adaptive_avg_pool1d           66          0        4    13.6s                                                                      
  nn.functional.adaptive_avg_pool2d           31          0        2     9.3s                                                                      
  nn.functional.grid_sample                   30          0        4   126.4s                                                                      
  nn.functional.interpolate                  280          0        2    83.9s                                                                      
  --------------------------------------------------------------------------                                                                       
  Total                                      622          0       ~43                                                                            
                                                                                                                                                   
  0 incorrect across the board. The remaining missing rules are:                                                                                 
                                                                                                                                                   
  - Upsample/interpolate ops (16): P(max)->P(max), P(min)->P(min) — invalid, interpolation blends values                                           
  - grid_sampler_2d (3-4): R, S(2)->S(3) (grid spatial sharding) + P(max/min) — invalid
  - _adaptive_avg_pool2d (2): P(max)->P(max), P(min)->P(min) — invalid                                                                             
  - mean.dim (4): P(avg)->P(avg), P(sum)->P(sum), S(2)->P(avg), S(3)->P(avg) — out of scope (decomposition chain)                                
  - ones_like (16): P(*)->P(*) / P(*)->S(3) — out of scope (decomposition chain)                                                                   
  - max_pool2d_with_indices_backward (28): mixed-Partial combos (e.g. P(max), P(avg), P(min) -> P(max)) — can't be expressed with current          
  framework, likely coincidental

```

### CODE REVIEW

File: torch/distributed/tensor/_ops/strategy_validation.py [hunk 1]
```diff
@@ -467,6 +470,14 @@ def validate_combination(
             device = tensors[0][1].device.type if tensors else "cpu"
             mesh = init_device_mesh(device, (world_size,))
 
+        # Uneven shards produce SymInt in LocalTensor's wrapper shape,
+        # which breaks C++ overload resolution before __torch_dispatch__
+        # can intercept. Return None to signal "untestable".
+        for (name, tensor), placement in zip(tensors, combination[0]):
```
Comment (Skylion007) [inline torch/distributed/tensor/_ops/strategy_validation.py:593]: Should zip be strict here?

File: test/distributed/tensor/test_dtensor_ops.py [hunk 2]
```diff
@@ -299,8 +306,14 @@ def repurpose_ops(op_db, base_test_name, derived_test_name):
     xfail("nn.functional.interpolate", "bicubic"),
     xfail("nn.functional.interpolate", "bilinear"),
     xfail("nn.functional.interpolate", "linear"),
+    xfail("nn.functional.interpolate", "nearest"),
+    xfail("nn.functional.interpolate", "nearest-exact"),
     xfail("nn.functional.interpolate", "trilinear"),
+    xfail("nn.functional.max_unpool1d"),
```
Comment (wconstab) [inline test/distributed/tensor/test_dtensor_ops.py:310]: why are so many xfails persisting after this PR? are some of these TODOs or are they all Won't Fix?  Probably add a comment for ones that are won't fix but for todos i wonder if this PR should address them or you have a good reason for deferring

File: torch/distributed/tensor/_ops/_math_ops.py [hunk 3]
```diff
@@ -1876,3 +1870,128 @@ def linalg_cross_strategy(
...
+    op: torch._ops.OpOverload,
+    args_schema: tuple[Any, ...],
+    kwargs_schema: dict[str, Any],
+) -> list[list[Placement | _ShardingPlaceholder]]:
+    # 1 output + 1 input = 2 placements; shard on batch (0) and channel (1)
+    # Upsample is a linear transformation so Partial(sum/avg) is valid.
+    return [
+        [_ShardingPlaceholder(0)] * 2,
+        [_ShardingPlaceholder(1)] * 2,
+        [Partial(), Partial()],
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/_math_ops.py]: is this shorthand for Partial("sum")?  prefer explicit

File: torch/distributed/tensor/_ops/_math_ops.py [hunk 4]
```diff
@@ -1876,3 +1870,128 @@ def linalg_cross_strategy(
...
+    # max_pool2d_with_indices_backward(grad_output, self, ..., indices) -> grad_input
+    # `self` is only used for output shape; `indices` drives the scatter so it
+    # must be identical on all ranks (Replicate).  The backward is a pure
+    # selection from grad_output, so Partial(sum/avg/max/min) all pass through.
+    input_strategy = cast(OpStrategy, op_schema.args_schema[0])
+    mesh = input_strategy.mesh
+    # n = 1 output + 3 tensor inputs
+    n = 4
+    single_mesh_dim_strategies: list[PlacementList] = [
+        [Replicate()] * n,
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/_math_ops.py]: usually we skip the Replicate strategy?

Comment (wconstab) [reply torch/distributed/tensor/_ops/_math_ops.py]: oh, this is not single-dim.  why not?

File: test/distributed/tensor/test_dtensor_ops.py (hunk 2 above)
Comment (zpcore) [reply test/distributed/tensor/test_dtensor_ops.py:310]: I think that's because those Ops now have the strategy implemented. So now we can see they begin fail in compile.

File: torch/distributed/tensor/_ops/_math_ops.py [hunk 5]
```diff
@@ -1881,3 +1875,123 @@ def linalg_cross_strategy(
...
+        [_ShardingPlaceholder(1)] * 2,
+        [Partial("sum"), Partial("sum")],
+        [Partial("avg"), Partial("avg")],
+    ]
+
+
+@register_single_dim_strategy(
+    [
+        aten.max_unpool2d.default,
+        aten.max_unpool3d.default,
```
Comment (zpcore) [inline torch/distributed/tensor/_ops/_math_ops.py:1923]: I think those two should allow [Partial(sum/avg), Partial(sum/avg), Replicate()].

**Review by zpcore — APPROVED** [review]
